### ISSUE TITLE
Feedback needed for OS X YARP Qt applications bundles

### ISSUE THREAD

**Issue by kt10aan** [robotology/yarp#340]
Since starting the qt apps from the terminal on OS X results in annoying behaviour (if you start e.g. the yarpmanager-qt from the terminal, this apps's menu doesn't show automatically, instead it is stuck on the Terminal.app's menu despite the yarpmanager window being on the foreground and it needs a switch back and forth to restore the yarpmanager-qt menu. Also, on quitting the app, things are getting slow), I ended up adding support for generating .app bundles for the Qt apps on OS X. (Manager, Logger, Scope).

It is quite simple. E.g. for Yarp Manager, in src/yarpmanager-qt/CMakeLists.txt I just add:

```
 if (APPLE)
      set(PROGNAME YarpManager)
      add_executable(${PROGNAME} MACOSX_BUNDLE ${yarpmanager_qt_SRCS}
                                      ${yarpmanager_qt_HDRS}
                                      ${yarpmanager_qt_QRC_GEN_SRCS}
                                      ${yarpmanager_qt_UI_GEN_SRCS})
                                  target_link_libraries(${PROGNAME} YARP_OS
                                       YARP_init
                                       YARP_manager)
                                   qt5_use_modules(${PROGNAME} Widgets
                                 Gui)
endif()
```

The produced YarpManager.app works nicely and behaves like a proper app, with no instabilities or interface glitches as the ones mentioned at the start of this post.

I will submit a PR for the Manager/Scope/Logger apps but before I do that I would like some feedback on the following questions:
- Is it ok to capitalise the apps' names as YarpManager/YarpLogger/YarpScope?
- I cannot seem to find any icons for these apps. Are they any icons so I can include them in the CMakeLists and have them automatically embedded in the apps (preferably HiRes!!)?
- Using the patch above, the usual creation of the executables (yarpmanager-qt/yarpmanager/yarplogger/yarpscope is not affected. The only thing that happens is that IF the computer runs OS X, _additional_ YarpXXX.apps are created and only for the ones that have a Qt GUI.
  Is this ok? Is it too much (given that if the system is OS X, the apps will automatically get created. I didn't want to bother people with extra CMake options)
- Should I also add an extra step that during make install the YarpXXX.apps are moved to the Applications folder or should I leave the user to do it manually?

**Comment by paulfitz** [comment]
Thanks for looking at this @kt10aan! Hopefully OSX users will chime in, but for what its worth here is some feedback from a just-uses-OSX-for-testing person:

 * Capitalization on .apps seems fine if this feels more "native" to you
 * I think there are some icons hidden in "icon_res.h" files in some of the gtk apps, not sure about qt.
 * The patch looks a little garbled towards the end? But adding .apps within an APPLE conditional seems very reasonable though :-)
 * About moving to Applications automatically on make install - I'd say follow prevailing practice (for minimal surprise to users), but I've no idea what that is.

**Comment by kt10aan** [comment]
* I am between Yarp Manager, YARPManager, Yarp Manager and YARP Manager styles. I will wait for OS X users input :)

* I found the hidden icons in the gtk apps but they are very low res. Are there any other icons around?

* The patch is just a replication of the previous (e.g.) yarpmanager executable calls, adding the OS X parameters

* Well, there is no specific prevailing practice, as compiling YARP is not a standard OS X application install anyway :) Homebrew leaves the generated .apps their own folders and then people can run an extra command to link/move them to /Applications

* The above reminded me that the yarp formula would have to be updated too, at least with a ```brew linkapps``` option :)

* FInally, another point I forgot. For proper/better/smoother bundling of the apps, a custom info.plist is needed to be put somewhere in the YARP source tree, each for each app. That info.plist file is copied into the .app. Is each app's respective folder a good option? (E.g. in src/yarpmanager-qt/Info.plist)?

**Comment by alecive** [comment]
@kt10aan Regarding the icons, I'm slowly uploading a better and more uniform iconset for them (only the qt version). I already did some work, but since it is quite a low priority thing, it's going to take a while.

I have uploaded the sources (`.svg` files) to a parallel repo (in order to not overwhelm the robotology one with useless files) with the icons I made up to now. They are available here https://github.com/alecive/iCubArtwork (under the `YARPGUIs` folder). @paulfitz @lornat75 if you think that they should go in the main repo, feel free to tell me.

I did not made any proper app icon (yet) because:
* I would like to have a better YARP logo first. From that one, I could draw any yarp-related icon. Btw my current tentative design is this one https://github.com/alecive/iCubArtwork/blob/master/YARPLogo/logoTestDifferentColors.png
* The gtk version did not allow any logo for how it was implemented
* I honestly don't know about the qt version

**Comment by kt10aan** [comment]
@drdanz, yes, strictly speaking its true, but there also other stuff that have to be done, like specifying icons, specifying plists with the $PATH and $YARP_DATA_DIRS, in the add_exectutable() call, (as the app won't inherit them from the command line) etc, so my thinking was that it would be more elegant and clear if that was done in a separate block of code for APPLE

**Comment by drdanz** [comment]
@kt10aan I agree that icons, plist, etc should be in a separate code block, but I don't think that having a different ``add_executable`` call would be a good idea... Mostly because it will cause the same executable to be built twice on APPLE
I believe that all the things that you need to set in the ``add_executable`` call can be set later as target properties...

**Comment by drdanz** [comment]
> specifying plists with the $PATH and $YARP_DATA_DIRS, in the add_exectutable() call

@kt10aan what do you mean? I don't see this part in the example...

**Comment by kt10aan** [comment]
@drdanz, the environmental variables  have to be passed to the Info.plist (in an LSEnvironment <dict> structure)  so that the resulting app can automatically find the xml files of yarp applications.

Yes, all the bundle customization stuff can happen in their own block (including the changing of the bundle name), but having the executable built two times (once as a console program and once as a bundle) is desired behaviour in my opinion. 

Mostly to minimise the impact on what the user is expecting from the build: The only difference would be the extra .app, the console program would still be there.
